File: semana_6_patrones_estructurales/codigo_ejemplos/ejemplo3_composite.cpp
/*
 * EJEMPLO 3: Composite en C++
 * (Ejemplo del PDF - Semana 6)
 */

#include <iostream>
#include <vector>

using namespace std;

class Figura {
public:
    virtual void dibujar() = 0;
};

class Circulo : public Figura {
public:
    void dibujar() override { 
        cout << "Circulo\n"; 
    }
};

class Grupo : public Figura {
    vector<Figura*> figuras;
public:
    void add(Figura* f) { 
        figuras.push_back(f); 
    }
    
    void dibujar() override {
        for (auto f : figuras) 
            f->dibujar();
    }
};

int main() {
    Circulo c1, c2;
    Grupo g;
    g.add(&c1);
    g.add(&c2);
    g.dibujar();
}

File: semana_6_patrones_estructurales/codigo_ejemplos/ejemplo3_composite.py
"""
EJEMPLO 3: Composite en Python
(Version cruzada del ejemplo C++ del PDF - Semana 6)
"""


class Figura:
    def dibujar(self):
        pass


class Circulo(Figura):
    def dibujar(self):
        print("Circulo")


class Grupo(Figura):
    def __init__(self):
        self.figuras = []
    
    def add(self, f):
        self.figuras.append(f)
    
    def dibujar(self):
        for f in self.figuras:
            f.dibujar()


c1 = Circulo()
c2 = Circulo()
g = Grupo()
g.add(c1)
g.add(c2)
g.dibujar()
